### PR TITLE
CompressLog: Add zstd compression

### PR DESCRIPTION
**Pull request by SuperSandro2000**
Basically #1219 rebased on master

- https://github.com/NixOS/nixpkgs/pull/333212

### CODE REVIEW

File: src/lib/Hydra/Plugin/CompressLog.pm [hunk 1]
```diff
@@ -9,11 +9,24 @@ use Hydra::Helper::CatalystUtils;
...
+    }
+
+    if ($silent ne '1') {
+        print STDERR "compressing '$logPath' with $compression...\n";
+    }
+
+    if ($compression eq 'bzip2') {
+        system('bzip2', '--force', $logPath);
+    } elsif ($compression eq 'zstd') {
+        system('zstd', '-T0', $logPath);
```
Comment (dasJ) [inline src/lib/Hydra/Plugin/CompressLog.pm:27]: Something that still annoys me with the old patch is that zstd is not quiet. Could you add this or do you think this adds any benefit?

**Comment by Mindavi** [comment]
See my comments @ nixpkgs.

File: src/lib/Hydra/Plugin/CompressLog.pm (hunk 1 above)
Comment (SuperSandro2000) [reply src/lib/Hydra/Plugin/CompressLog.pm:27]: I also disabled it in https://github.com/NixOS/nixpkgs/pull/332752, we don't need three lines of useless progress bars.

File: src/lib/Hydra/Plugin/CompressLog.pm [hunk 2]
```diff
@@ -9,11 +9,24 @@ use Hydra::Helper::CatalystUtils;
...
+    }
+
+    if ($silent ne '1') {
+        print STDERR "compressing '$logPath' with $compression...\n";
+    }
+
+    if ($compression eq 'bzip2') {
+        system('bzip2', '--force', $logPath);
+    } elsif ($compression eq 'zstd') {
+        system('zstd', '--quiet', -T0', $logPath);
```
Comment (Mindavi) [inline src/lib/Hydra/Plugin/CompressLog.pm:27]:
```suggestion
        system('zstd', '--quiet', '-T0', $logPath);
```

**Comment by Mindavi** [comment]
Looks good to me, but haven't tested. Would really love it if a small test is added, but maybe that's a big ask. Would definitely improve my confidence, at least of the case where CompressLog is called. The fallback is probably a bit more annoying to test.

File: src/lib/Hydra/Plugin/CompressLog.pm [hunk 3]
```diff
@@ -9,11 +9,24 @@ use Hydra::Helper::CatalystUtils;
...
+    }
+
+    if ($silent ne '1') {
+        print STDERR "compressing '$logPath' with $compression...\n";
+    }
+
+    if ($compression eq 'bzip2') {
+        system('bzip2', '--force', $logPath);
+    } elsif ($compression eq 'zstd') {
+        system('zstd', '--quiet', '-T0', $logPath);
```
Comment (SuperSandro2000) [inline src/lib/Hydra/Plugin/CompressLog.pm:27]:
```suggestion
        system('zstd', '--rm', '--quiet', '-T0', $logPath);
```

if we want to match bzip2, we need to rm the files.

**Comment by SuperSandro2000** [comment]
> Would really love it if a small test is added, but maybe that's a big ask.

I have tested this on my hydra and from what I can tell, it seems to work minus the bit that the original logs where not removed. I don't have time to write a test especially with the current test infrastructure.

**Comment by SuperSandro2000** [comment]
I did some more testing and the logs where not displayed in the web ui. That is now also fixed. I think I have now tested this change end to end and hopefully found all the misses along the way.

**Review by Mindavi — APPROVED** [review]
Diff now looks good to me, unfortunately I cannot test it this week. Maybe next week.

Out of curiosity, what happens if you fill in something else than bz2 or zstd? It'll just fail somewhere in the compression routines?

File: nixos-modules/hydra.nix [hunk 4]
```diff
@@ -458,10 +459,15 @@ in
...
     systemd.services.hydra-compress-logs =
-      { path = [ pkgs.bzip2 ];
+      { path = [ pkgs.bzip2 pkgs.zstd ];
         script =
           ''
-            find ${baseDir}/build-logs -type f -name "*.drv" -mtime +3 -size +0c | xargs -r bzip2 -v -f
+            set -eou pipefail
+            compression=$(sed -nr 's/compress_build_logs_compression = ()/\1/p' ${baseDir}/hydra.conf)
+            if [[ $compression == zstd ]]; then
+              compression=zstd --rm
```
Comment (SuperSandro2000) [inline nixos-modules/hydra.nix:468]:
```suggestion
              compression="zstd --rm"
```

**Comment by SuperSandro2000** [comment]
> Out of curiosity, what happens if you fill in something else than bz2 or zstd? It'll just fail somewhere in the compression routines?

I hope so. 😅 I didn't test this explicitly.

**Review by dasJ — APPROVED** [review]

**Pull request merged**